### PR TITLE
[WIP] Replace torchaudio rnnt_loss with k2 pruned rnnt loss

### PR DESCRIPTION
**Pull request by pkufool**
After a month of experiments，I finally manage to make the pruned rnnt loss work. Thank for Daniel's help! 

By using the pruned version loss, we can train with larger batch size (max-duration = 300), and it only costs 2 hours to finish one epoch on full librispeech dataset (using 4 GPUs). Below are some experiment results, the WER is little worse than the model trained with torch rnnt loss, we will continue tuning the model to make it better.

For more details of the pruned rnnt loss, please see https://github.com/k2-fsa/k2/pull/891.

best WER for (greedy search) | baseline (trained with torch rnnt loss) <br> max-duration = 200 <br> 4 GPUs <br> test-clean \|\| test-other | k2 pruned rnnt loss (range = 3)<br> max-duration = 300 <br> 3 GPUs <br> test-clean \|\| test-other   | k2 pruned rnnt loss (range = 5) <br> max-duration = 300 <br> 4 GPUs<br> test-clean \|\| test-other  | k2 pruned rnnt loss (range = 8) <br> max-duration = 300 <br> 4 GPUs<br> test-clean \|\| test-other 
-- | -- | -- | -- | --
epoch 4 |   | 5.86 \|\| 14.13 | 6.75 \|\| 15.91 | 6.72 \|\| 15.34
epoch 5 |   | 5.09 \|\| 12.57 | 5.6 \|\| 13.65 | 5.59 \|\| 13.52
epoch 6 | 4.24 \|\| 10.6 | 4.83 \|\| 11.87 | 4.99 \|\| 12.38 | 5.07 \|\| 12.23
epoch 7 | 3.95 \|\| 9.87 |   | 4.65 \|\| 11.3 | 4.46 \|\| 11.5
epoch 8 | 3.64 \|\| 9.29 | 4.11 \|\| 9.23 | 4.38 \|\| 10.58 | 4.34 \|\| 10.53
epoch 9 | 3.57 \|\| 8.99 |   | 4.12 \|\| 10.02 | 4.06 \|\| 10.14
epoch 10 | 3.49 \|\| 8.69 | 3.65 \|\| 9.23 | 3.9 \|\| 9.66 | 3.82 \|\| 9.66
epoch 11 |   |   |   |  
epoch 12 | 3.24 \|\| 8.17 | 3.38 \|\| 8.56 | 3.61 \|\| 9.01 | 3.53 \|\| 8.89
epoch 14 | 3.16 \|\| 8.05 | 3.31 \|\| 8.4 | 3.46 \|\| 8.59 | 3.34 \|\| 8.54
epoch 15 | 3.12 \|\| 7.95 |   | 3.43 \|\| 8.41 | 3.32 \|\| 8.35
epoch 16 | 3.06 \|\| 7.99 | 3.27 \|\| 8.05 | 3.4 \|\| 8.31 | 3.29 \|\| 8.23
epoch 17 | 3.07 \|\| 7.86 |   | 3.35 \|\| 8.12 | 3.22 \|\| 8.02
epoch 18 | 3.02 \|\| 7.76 | 3.24 \|\| 7.91 | 3.31 \|\| 8.01 | 3.19 \|\| 8.0
epoch 19 | 3.0 \|\| 7.69 |   | 3.26 \|\| 7.93 | 3.15 \|\| 7.82
epoch 20 | 2.99 \|\| 7.61 | 3.14 \|\| 7.76 | 3.21 \|\| 7.78 | 3.09 \|\| 7.82
epoch 21 | 2.98 \|\| 7.51 | 3.11 \|\| 7.75 |   |  
epoch 22 | 2.95 \|\| 7.5 |   | 3.17 \|\| 7.97 | 3.0 \|\| 7.67
epoch 23 | 2.93 \|\| 7.43 | 3.1 \|\| 7.64 |   | 3.0 \|\| 7.66
epoch 24 | 2.92 \|\| 7.38 |   | 3.11 \|\| 7.68 | 2.98 \|\| 7.59
epoch 25 | 2.91 \|\| 7.37 |   |   |  
epoch 26 | 2.9 \|\| 7.34 |   | 3.06 \|\| 7.61 | 2.98 \|\| 7.47
epoch 27 | 2.87 \|\| 7.29 |   | 3.05 \|\| 7.55 | 3.0 \|\| 7.55
epoch 28 | 2.87 \|\| 7.27 |   | 3.05 \|\| 7.69 | 2.97 \|\| 7.43
epoch 29 | 2.85 \|\| 7.28 |   | 3.0 \|\| 7.47 | 2.98 \|\| 7.39
epoch 30 | 2.85 \|\| 7.29 |   |   |  
epoch 31 | 2.85 \|\| 7.26 |   |   | 2.94 \|\| 7.33
epoch 32 | 2.85 \|\| 7.27 |   | 2.98 \|\| 7.29 | 2.94 \|\| 7.33
epoch 33 | 2.85 \|\| 7.3 |   |   |

### CODE REVIEW

**Comment by pkufool** [comment]
The future experiments:

- [ ] try using rnnt_loss_simple as a regularization item, (i.e., loss = small_scale * simple_loss + pruned_loss)
- [ ] try the rnnt_loss_smoothed using lm_only & am_only options
- [ ] try training with float16.

Will update experiment results here once they are available.

**Comment by danpovey** [comment]
Great work!!
I think the easiest way to get the same WERs as Fangjun is to train for a
big longer or maybe use a slightly larger
learning-rate factor.
Likely the higher max-duration makes it train slower.

**Comment by pkufool** [comment]
Thanks, will continue training the model and also try larger lr-factor.

**Comment by pkufool** [comment]
The results of **using lm_only/am_only smoothing** are shown in the table below, the second column is the result of baseline model trained with pytorch rnnt_loss, the third column is the result of the model trained with k2 rnnt_loss_pruned, the forth column is the result of the model trained with k2 rnnt_loss_pruned with `lm_only_scale=0.25`,  and the last column is the result of the model trained with k2 rnnt_loss_pruned with `am_only_scale=0.25`.  Adding lm_only/am_only smoothing may have other benefits, as for the effects on WERs, you can see in the table, `lm_only_scale` option helps to converge faster and get better results.


best WER for (greedy search) | baseline (torch rnnt loss) test-clean \|\| test-other max-duration = 200 4 GPUs | (k2 pruned loss range 8) test-clean \|\| test-other max-duration = 300 4 GPUs | (k2 pruned loss range 8) lm_only_scale=0.25 am_only_scale=0.0 test-clean \|\| test-other max-duration = 300 4 GPUs | (k2 pruned loss range 8) lm_only_scale=0.0 am_only_scale=0.25 test-clean \|\| test-other max-duration = 300 4 GPUs
-- | -- | -- | -- | --
epoch 10 | 3.49 \|\| 8.69 | 3.82 \|\| 9.66 | 3.92 \|\| 9.96 | 3.96 \|\| 9.73
epoch 11 |   |   | 3.73 \|\| 9.34 | 3.76 \|\| 9.3
epoch 12 | 3.24 \|\| 8.17 | 3.53 \|\| 8.89 | 3.52 \|\| 9.0 | 3.68 \|\| 8.93
epoch 13 |   |   | 3.45 \|\| 8.73 | 3.55 \|\| 8.72
epoch 14 | 3.16 \|\| 8.05 | 3.34 \|\| 8.54 | 3.35 \|\| 8.47 | 3.49 \|\| 8.62
epoch 15 | 3.12 \|\| 7.95 | 3.32 \|\| 8.35 | 3.19 \|\| 8.28 | 3.4 \|\| 8.38
epoch 16 | 3.06 \|\| 7.99 | 3.29 \|\| 8.23 | 3.14 \|\| 8.13 | 3.33 \|\| 8.29
epoch 17 | 3.07 \|\| 7.86 | 3.22 \|\| 8.02 | 3.05 \|\| 8.03 | 3.26 \|\| 8.27
epoch 18 | 3.02 \|\| 7.76 | 3.19 \|\| 8.0 | 3.06 \|\| 7.95 | 3.22 \|\| 8.14
epoch 19 | 3.0 \|\| 7.69 | 3.15 \|\| 7.82 | 3.01 \|\| 7.79 | 3.18 \|\| 7.94
epoch 20 | 2.99 \|\| 7.61 | 3.09 \|\| 7.82 | 2.97 \|\| 7.72 | 3.17 \|\| 7.85
epoch 21 | 2.98 \|\| 7.51 |   | 3.0 \|\| 7.73 | 3.16 \|\| 7.72
epoch 22 | 2.95 \|\| 7.5 | 3.0 \|\| 7.67 | 2.99 \|\| 7.59 | 3.11 \|\| 7.71
epoch 23 | 2.93 \|\| 7.43 | 3.0 \|\| 7.66 | 2.95 \|\| 7.59 | 3.1 \|\| 7.66
epoch 24 | 2.92 \|\| 7.38 | 2.98 \|\| 7.59 | 2.96 \|\| 7.5 | 3.06 \|\| 7.71
epoch 25 | 2.91 \|\| 7.37 |   | 2.97 \|\| 7.37 | 3.02 \|\| 7.69
epoch 26 | 2.9 \|\| 7.34 | 2.98 \|\| 7.47 | 2.92 \|\| 7.33 | 3.01 \|\| 7.62
epoch 27 | 2.87 \|\| 7.29 | 3.0 \|\| 7.55 | 2.92 \|\| 7.26 | 3.01 \|\| 7.56
epoch 28 | 2.87 \|\| 7.27 | 2.97 \|\| 7.43 | 2.89 \|\| 7.18 | 3.04 \|\| 7.57
epoch 29 | 2.85 \|\| 7.28 | 2.98 \|\| 7.39 | 2.88 \|\| 7.22 | 3.01 \|\| 7.57
epoch 30 | 2.85 \|\| 7.29 |   | 2.89 \|\| 7.07 | 3.01 \|\| 7.59
epoch 31 | 2.85 \|\| 7.26 | 2.94 \|\| 7.33 | 2.89 \|\| 6.97 | 3.01 \|\| 7.54
epoch 32 | 2.85 \|\| 7.27 | 2.94 \|\| 7.33 | 2.86 \|\| 6.94 | 2.98 \|\| 7.51

**Comment by pkufool** [comment]
> I think the easiest way to get the same WERs as Fangjun is to train for a big longer or maybe use a slightly larger learning-rate factor. Likely the higher max-duration makes it train slower.

After training with more epochs, the model trained with k2 rnnt_loss can get same WERs as pytorch baseline (i.e. 2.85 || 7.26) at epoch 44. Will try with larger lr-factor to see if the model could converge faster.

**Comment by danpovey** [comment]
Great!!
I suppose you can try different values of lmonly_scale, like 0.1 or 0.5.  It may not be very sensitive to that, is my guess,
but worth finding out.
And you could also see whether the am_only_scale experiment behaves differently with the max_symbols_per_frame
in decoding.
Something else worth looking into, is the overall scale on the "simple" version of the RNN-T loss.  I believe it's 1.0
now, could try 0.5 for instance.  It would be nice if you could let us know, in some compact way (those TF logs are
hard for me to view), whether the pruned loss seems to overfit less than the standard loss.  I would expect that
it would, i.e. train/valid difference should be a bit less, because it's regularized by the "simple" loss that has the trivial joiner.

**Comment by pkufool** [comment]
> It would be nice if you could let us know, in some compact way (those TF logs are hard for me to view), whether the pruned loss seems to overfit less than the standard loss. I would expect that it would, i.e. train/valid difference should be a bit less, because it's regularized by the "simple" loss that has the trivial joiner.

I extracted the train/valid loss from the logs, pictures below are the difference of train/valid difference for different loss. It seems that the **am/lm smoothing** do help to avoid overfitting.

| loss | train/valid loss curve|  (train - val) / train curve|
| -----| -----| -----|
| Unpruned loss|image| image|
| Pruned loss (s_range=8)|image|image|
| Pruned loss with am smoothing (s_range=8)|image| image|
| Pruned loss with lm smoothing (s_range=8)| image| image|

**Comment by danpovey** [comment]
Are all those plots for the same number of epochs?
I find these numbers a little hard to understand.  
Firstly, I would have expected the "simple loss" that we use for pruning would have the main regularization effect, but that's now what see.  And I wasn't really expecting the am and lm smoothing to have much of an effect on the train/valid difference.
Are these for the same number of epochs, and were they with similar settings w.r.t. learning rate and number of workers?

**Comment by pkufool** [comment]
> Are all those plots for the same number of epochs? 

The unpruned loss curve contains 46 epochs, and the other three contains 49 epochs.


I find these numbers a little hard to understand. Firstly, I would have expected the "simple loss" that we use for pruning would have the main regularization effect, but that's not what we see. And I wasn't really expecting the am and lm smoothing to have much of an effect on the train/valid difference. Are these for the same number of epochs, and were they with similar settings w.r.t. learning rate and number of workers?

**Comment by pkufool** [comment]
> Are all those plots for the same number of epochs? 

Unpruned curve plots 46 epochs, the other three plots 49 epochs.

> were they with similar settings w.r.t. learning rate and number of workers?

The unpruned one trained with 3 GPUs with max-duration=250, the other three trained with 4 GPUs with max-duration=300, the lr-factor and warmup-steps are all same.

> [edit]: I suspect these are not just the pruned part of the loss that you are plotting; if they were just the pruned loss, they'd all have about the same magnitude. I think you are plotting the total loss including the simple loss, which is not really comparable across the different experiments because they are different loss functions.

The pictures below show the difference with only pruned loss.

| loss | train/valid loss curve|  (train - val) / train curve|
| -----| -----| -----|
| Unpruned loss|image| image|
| Pruned loss (s_range=8)|image|image|
| Pruned loss with am smoothing (s_range=8)|image|image|
| Pruned loss with lm smoothing (s_range=8)|image|image|

**Comment by danpovey** [comment]
OK.. so the only clear improvement in overfitting is with the "am loss", which anyway was not working great WER-wise...
possibly that one's just underfitting.

**Comment by pkufool** [comment]
> > I think the easiest way to get the same WERs as Fangjun is to train for a big longer or maybe use a slightly larger learning-rate factor. Likely the higher max-duration makes it train slower.
> 
> After training with more epochs, the model trained with k2 rnnt_loss can get same WERs as pytorch baseline (i.e. 2.85 || 7.26) at epoch 44. Will try with larger lr-factor to see if the model could converge faster.

I tried to use a slightly larger lr-factor ( 2.5 -> 5), the model can coverge faster.  Now we can get the same WERs with same epochs as baseline. And it takes less time for our setting (2 hours per epoch vs more than 3 hours per epoch).

|  Best WER for epochs (greedy search) | baseline using pytorch loss, max-duration=200 | k2 pruned loss (s_range=5), max-duration=300, lr-factor=2.5| k2_pruned_loss (s_range=5), max-duration=300, lr-factor=5|
|---|---|---|---|
|epoch 30 |2.85 \|\| 7.26 | |2.85 \|\| 7.03 |
|epoch 31 | 2.85 \|\| 7.27| 3.02 \|\|7.42  | 2.84 \|\| 7.02 |
|epoch 32|2.85 \|\| 7.3 | 2.98 \|\| 7.29 | 2.82 \|\| 7.01 |

**Comment by pkufool** [comment]
@danpovey  I just noticed that I made a small mistake about the pruned loss training, thanks to @pingfengluo . We need the am & lm tensor to get the `s_begin` (in rnnt_loss_simple), so the shape of am should be [B, T, vocab-size] and lm [B, S+1, vocab-size]. The default setting of the output shape of am&lm is [..., 512] and I used a bpe vocab-size of 500, so it did not raise an error.  Hope this would not hurt the WER.

**Comment by danpovey** [comment]
It's OK, should not make a difference to WER as extra rows would be ignored.

**Comment by pkufool** [comment]
See updates in #194
